mlir/lib/Transforms/lab4/nogin_denis/FusedMultAddPass.cpp

```diff
@@ -0,0 +1,57 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class FusedMultAddPass
+    : public PassWrapper<FusedMultAddPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final { return "fused-mult-add"; }
+  StringRef getDescription() const final {
+    return "Fuses multiply and add operations into a single fma operation";
+  }
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    module.walk([&](Operation *op) {
+      if (auto addOp = dyn_cast<LLVM::FAddOp>(op)) {
+        Value addLHS = addOp.getOperand(0);
+        Value addRHS = addOp.getOperand(1);
+
+        if (auto mulOpLHS = addLHS.getDefiningOp<LLVM::FMulOp>()) {
+          OpBuilder builder(addOp);
+          Value fma = builder.create<LLVM::FMAOp>(
+              addOp.getLoc(), mulOpLHS.getOperand(0), mulOpLHS.getOperand(1),
+              addRHS);
+          addOp.replaceAllUsesWith(fma);
+          addOp.erase();
+          mulOpLHS.erase();
```

Comment:
Actually, mulOp may have other uses except addOp. You need to check that

Reply:
Also test for that should be added

Reply:
Please explain what you mean. What other uses can we use mulOp for?

Reply:
You can take a look here: https://github.com/NN-complr-tech/llvm-nnsu-2024/pull/152#discussion_r1605668496